contrib/runtime/aer_runtime.cpp

```diff
@@ -0,0 +1,216 @@
+  AER::AerState* state = reinterpret_cast<AER::AerState*>(handler);
+  state->apply_mcrz({qubit}, theta);
+};
+
+// controlled-NOT
+void aer_apply_cx(void* handler, uint_t ctrl_qubit, uint_t tgt_qubit) {
+  AER::AerState* state = reinterpret_cast<AER::AerState*>(handler);
+  state->apply_mcx({ctrl_qubit, tgt_qubit});
+};
+
+// controlled-Y
+void aer_apply_cy(void* handler, uint_t ctrl_qubit, uint_t tgt_qubit) {
+  AER::AerState* state = reinterpret_cast<AER::AerState*>(handler);
+  state->apply_mcy({ctrl_qubit, tgt_qubit});
+};
+
+// controlled-Z
+void aer_apply_cz(void* handler, uint_t ctrl_qubit, uint_t tgt_qubit) {
+  AER::AerState* state = reinterpret_cast<AER::AerState*>(handler);
+  state->apply_mcz({ctrl_qubit, tgt_qubit});
+};
+
+// controlled-phase
+void aer_apply_cp(void* handler, uint_t ctrl_qubit, uint_t tgt_qubit, double lambda) {
+  AER::AerState* state = reinterpret_cast<AER::AerState*>(handler);
+  state->apply_mcphase({ctrl_qubit, tgt_qubit}, lambda);
+};
+
+// controlled-rx
+void aer_apply_crx(void* handler, uint_t ctrl_qubit, uint_t tgt_qubit, double theta) {
+  AER::AerState* state = reinterpret_cast<AER::AerState*>(handler);
+  state->apply_mcrx({ctrl_qubit, tgt_qubit}, theta);
+};
+
+// controlled-ry
+void aer_apply_cry(void* handler, uint_t ctrl_qubit, uint_t tgt_qubit, double theta) {
+  AER::AerState* state = reinterpret_cast<AER::AerState*>(handler);
+  state->apply_mcry({ctrl_qubit, tgt_qubit}, theta);
+};
+
+// controlled-rz
+void aer_apply_crz(void* handler, uint_t ctrl_qubit, uint_t tgt_qubit, double theta) {
+  AER::AerState* state = reinterpret_cast<AER::AerState*>(handler);
+  state->apply_mcrz({ctrl_qubit, tgt_qubit}, theta);
+};
+
+// controlled-H
+void aer_apply_ch(void* handler, uint_t ctrl_qubit, uint_t tgt_qubit) {
+  AER::AerState* state = reinterpret_cast<AER::AerState*>(handler);
+  state->apply_mcu({ctrl_qubit, tgt_qubit}, M_PI / 2.0, 0, M_PI);
+};
+
+// swap
+void aer_apply_swap(void* handler, uint_t qubit0, uint_t qubit1) {
+  AER::AerState* state = reinterpret_cast<AER::AerState*>(handler);
+  state->apply_mcswap({qubit0, qubit1});
+};
+
+// Toffoli
+void aer_apply_ccx(void* handler, uint_t qubit0, uint_t qubit1, uint_t qubit2) {
+  AER::AerState* state = reinterpret_cast<AER::AerState*>(handler);
+  state->apply_mcx({qubit0, qubit1, qubit2});
+};
+
+// // controlled-swap
+void aer_apply_cswap(void* handler, uint_t ctrl_qubit, uint_t qubit0, uint_t qubit1) {
+  AER::AerState* state = reinterpret_cast<AER::AerState*>(handler);
+  state->apply_mcswap({ctrl_qubit, qubit0, qubit1});
+};
+
+// four parameter controlled-U gate with relative phase γ
+void aer_apply_cu(void* handler, uint_t ctrl_qubit, uint_t tgt_qubit, double theta, double phi, double lambda, double gamma) {
+  AER::AerState* state = reinterpret_cast<AER::AerState*>(handler);
+  state->apply_mcphase({ctrl_qubit}, gamma);
+  state->apply_mcu({ctrl_qubit, tgt_qubit}, theta, phi, lambda);
+};
+}
```

Comment:
Please let me leave unitary as follow up. memory alignment of a complex matrix is different in applications. I would like to follow main usecase.

Reply:
Yes, that is fine, let's do it in a follow up later. I only asked because it was the only obvious thing I didn't see. I can see how it would be more complicated though.